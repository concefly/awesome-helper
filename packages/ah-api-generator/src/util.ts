import { OpenAPIV3 as API } from 'openapi-types';
import _ from 'lodash';
import { Schema } from 'jsonschema';

import { inspect } from 'util';

export const log = (...args: any[]) => {
  console.log(...args.map(arg => inspect(arg, { showHidden: false, depth: null })));
};

export function parseRequestSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  if (!data) return;

  const schema: Schema = {
    type: 'object',
    properties: {},
  };

  if (data.parameters) {
    (data.parameters as API.ParameterObject[]).forEach(p => {
      Object.assign(schema.properties!, {
        [p.name]: p.schema,
      });
    });
  }

  if (data.requestBody) {
    const requestBodyContentJson = (data.requestBody as API.RequestBodyObject).content[
      'application/json'
    ];

    if (typeof (requestBodyContentJson?.schema as any)?.type !== 'undefined') {
      const jSchema = requestBodyContentJson?.schema as Schema;
      _.merge(schema, jSchema);
    }
  }

  // 没有收集到，返回空
  if (_.isEmpty(schema.properties)) return;

  return schema;
}

export function parseResponseSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  const response = data?.responses?.['200'] as API.ResponseObject;
  const jsonResponse = response.content?.['application/json'];

  if (jsonResponse) {
    // 尝试取 schema
    if (jsonResponse.schema) return jsonResponse.schema as Schema;

    const example =
      jsonResponse.example || (jsonResponse.examples && Object.values(jsonResponse.examples).pop());

    // 尝试转换 example
    if (example) return data2Schema(example);
  }
}

export type IRequestMeta = {
  query?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  response?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  pathName: string;
  method: string;
  description?: string;
};

export interface IServiceMap {
  [serviceName: string]: {
    [requestName: string]: IRequestMeta;
  };
}

export function renderService(apiDoc: API.Document, opt: { operationIdSplitter?: string } = {}) {
  const serviceMap: IServiceMap = {};
  const { operationIdSplitter = '.' } = opt;

  Object.entries(apiDoc.paths).forEach(([pathName, pi]) => {
    if (!pi) return;

    Object.entries({
      GET: pi.get,
      PUT: pi.put,
      POST: pi.post,
      DELETE: pi.delete,
    }).forEach(([method, commonPiData]) => {
      if (!commonPiData) return;

      let [serviceName, requestName] = (
        commonPiData.operationId || `default${operationIdSplitter}${_.camelCase(pathName)}`
      ).split(operationIdSplitter);

      serviceName = _.capitalize(serviceName);

      const requestSchema = parseRequestSchema(commonPiData);
      const respSchema = parseResponseSchema(commonPiData);

      const requestMeta: IRequestMeta = {
        query: requestSchema
          ? { schema: requestSchema, tsTypeLiteral: schema2TsTypeLiteral(requestSchema) }
          : undefined,
        response: respSchema
          ? { schema: respSchema, tsTypeLiteral: schema2TsTypeLiteral(respSchema) }
          : undefined,
        method,
        pathName,
        description: commonPiData.description,
      };

      //
      _.set(serviceMap, [serviceName, requestName], requestMeta);
    });
  });

  return serviceMap;
}

export function schema2TsTypeLiteral(s?: Schema): string {
  if (!s) return 'never';

  if (s.type === 'string') return 'string';
  if (s.type === 'integer') return 'number';

  if (s.type === 'array') {
    return `Array<${schema2TsTypeLiteral((s.items as any) || '{}')}>`;
  }

  if (s.type === 'object') {
    {
      if (s.properties) {
        return [
          '{',
          Object.entries(s.properties).map(([pn, pv]) => {
            const isRequired =
              typeof s.required === 'boolean'
                ? s.required
                : Array.isArray(s.required)
                ? s.required.includes(pn)
                : false;

            const tsType = schema2TsTypeLiteral(pv);

            return `${pn}${isRequired ? '' : '?'}: ${tsType}`;
          }),
          '}',
        ].join(' ');
      }

      return 'object';
    }
  }

  return 'any';
}

/** 从数据结构猜测 schema */
export function data2Schema(data: any): Schema {
  if (_.isString(data)) return { type: 'string' };
  if (_.isInteger(data)) return { type: 'integer' };
  if (_.isNumber(data)) return { type: 'number' };

  if (_.isBoolean(data)) return { type: 'boolean' };

  if (_.isArray(data)) {
    return {
      type: 'array',
      items: data2Schema(data[0]),
    };
  }

  if (_.isPlainObject(data)) {
    return {
      type: 'object',
      properties: _.mapValues(data, data2Schema),
      required: Object.keys(data),
    };
  }

  return {};
}
