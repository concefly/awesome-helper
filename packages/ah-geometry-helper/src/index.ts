export * from './IDirType';
export * from './Line';
export * from './LineSegment';
export * from './LineSet';
export * from './Point';
export * from './Rect';
