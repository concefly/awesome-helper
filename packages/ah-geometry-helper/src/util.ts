export function inRange(x: number, a: number, b: number) {
  return a <= x && x <= b;
}

export function minBy<T>(list: T[], mapper: (d: T) => number): T {
  const tempMap = new Map<number, T>();

  const mapValues = list.map(d => {
    const n = mapper(d);
    tempMap.set(n, d);

    return n;
  });

  return tempMap.get(Math.min(...mapValues))!;
}
