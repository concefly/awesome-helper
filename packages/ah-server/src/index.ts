import 'koa';
import { DefaultContext, DefaultState, ParameterizedContext } from 'koa';

export interface IApplication {}
export interface IService {}

export interface IContext extends ParameterizedContext<DefaultState, DefaultContext> {}
export interface IConfig {
  LOCAL_PORT: number;
}

// fix koa plugin type
declare module 'koa' {
  interface Request {
    // koaBody type
    body?: any;
  }
}

export * from './App';
export * from './Config';
export * from './Logger';
export * from './Controller';
export * from './Service';
export * from './util';
export * from './Scheduler';
export * from './error';
