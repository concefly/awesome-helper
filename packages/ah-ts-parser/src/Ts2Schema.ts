import { Schema } from 'jsonschema';
import { Project, SourceFile, Type, Symbol, Node } from 'ts-morph';

function symbol2Type(sym: Symbol) {
  return sym.getTypeAtLocation(sym.getValueDeclarationOrThrow());
}

export class Ts2Schema {
  constructor(readonly tsConfigFilePath: string) {}

  private proj = new Project({ tsConfigFilePath: this.tsConfigFilePath });

  private type2Schema(type: Type, depth = 0): Schema {
    if (depth >= 9) return {};

    const title = type.getSymbol()?.getEscapedName();

    if (type.isString() || type.isStringLiteral()) return { title, type: 'string' };
    if (type.isNumber() || type.isNumberLiteral()) return { title, type: 'integer' };

    // class
    if (type.isClass()) {
      const sym = type.getSymbolOrThrow();
      const properties: any = {};

      sym.getMembers().forEach(mSym => {
        const mType = symbol2Type(mSym);

        properties[mSym.getEscapedName()] = this.type2Schema(mType, depth + 1);
      });

      return {
        title,
        type: 'object',
        properties,
      };
    }

    const callSignatures = type.getCallSignatures();
    if (callSignatures.length) {
      const fNode = type.getSymbolOrThrow().getValueDeclarationOrThrow();
      let decSchema!: Schema;

      if (Node.isMethodDeclaration(fNode)) {
        const decorators = fNode.getDecorators();
        if (decorators.length) {
          decSchema = {
            type: 'array',
            items: decorators.map(d => {
              const decExp = d.getCallExpressionOrThrow();

              return {
                title: d.getName(),
                type: 'object',
                properties: {
                  arguments: {
                    type: 'array',
                    items: decExp.getArguments().map(argNode => {
                      if (Node.isStringLiteral(argNode)) {
                        return { type: 'string', const: argNode.getText() } as Schema;
                      }
                      return {};
                    }),
                  },
                },
              } as Schema;
            }),
          };
        }
      }

      return {
        title,
        type: 'object',
        format: 'function',
        properties: {
          ...(decSchema && { decorators: decSchema }),
          signatures: {
            type: 'array',
            items: callSignatures.map(callSig => {
              return {
                type: 'object',
                properties: {
                  params: {
                    type: 'array',
                    items: callSig
                      .getParameters()
                      .map(p => symbol2Type(p))
                      .map(t => this.type2Schema(t, depth + 1)),
                  },
                  return: this.type2Schema(callSig.getReturnType(), depth + 1),
                },
              } as Schema;
            }),
          },
        },
      };
    }

    // promise
    if (type.getProperty('then') && type.getProperty('catch')) {
      return {
        type: 'object',
        format: 'promise',
        properties: {
          typeArguments: {
            type: 'array',
            items: type.getTypeArguments().map(argType => this.type2Schema(argType, depth + 1)),
          },
        },
      };
    }

    // any object
    if (type.isObject()) {
      const properties: any = {};

      type.getApparentProperties().forEach(ap => {
        const apt = ap.getTypeAtLocation(ap.getValueDeclarationOrThrow());
        properties[ap.getEscapedName()] = this.type2Schema(apt, depth + 1);
      });

      return {
        type: 'object',
        properties,
      };
    }

    return {};
  }

  private getSchemaOneFile(file: SourceFile) {
    const schema: Schema = {
      title: file.getFilePath().replace(process.cwd() + '/', ''),
      type: 'object',
      properties: {
        classes: {
          type: 'array',
          items: file.getClasses().map(cls => this.type2Schema(cls.getType())),
        },
        functions: {
          type: 'array',
          items: file.getFunctions().map(fn => this.type2Schema(fn.getType())),
        },
      },
    };

    return schema;
  }

  getSchema(pattern: string) {
    const files = this.proj.getSourceFiles(pattern);

    const schema: Schema = {
      type: 'object',
      properties: {
        files: {
          type: 'array',
          items: files.map(file => this.getSchemaOneFile(file)),
        },
      },
    };

    return schema;
  }
}
