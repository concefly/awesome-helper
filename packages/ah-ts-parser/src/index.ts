export * from 'ts-morph';
export * from './Ts2Schema';
export * from './util';
