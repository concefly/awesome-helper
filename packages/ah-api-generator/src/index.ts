import path from 'path';
import fs from 'fs';
import { curl } from 'urllib';
import { renderService } from './util';
import ejs from 'ejs';
import { OpenAPIV3 as API } from 'openapi-types';
import prettier from 'prettier';
import RefParser from '@apidevtools/json-schema-ref-parser';

export interface IConfig {
  input?: { type: 'local'; filename: string } | { type: 'remote'; url: string; headers?: any };
  operationIdSplitter?: string;
  template?: string;
  banner?: string;
  dump?: string;
}

export const loadApiDoc = async (input: IConfig['input']): Promise<API.Document> => {
  if (!input) throw new Error('input 不存');

  let result =
    input.type === 'local'
      ? JSON.parse(fs.readFileSync(input.filename, { encoding: 'utf-8' }))
      : input.type === 'remote'
      ? await curl<any>(input.url, { headers: input.headers, dataType: 'json' }).then(r => r.data)
      : null;

  if (!result) throw new Error(`inputSource 不存在: ${JSON.stringify(input)}`);

  // 解 $ref
  result = await RefParser.dereference(result);

  return result;
};

export const generateAPIService = async (cfg: IConfig) => {
  const {
    input = { type: 'local', filename: 'api.json' },
    template = path.resolve(__dirname, '..', 'template.ejs'),
    dump,
    banner,
  } = cfg;

  const apiDoc = await loadApiDoc(input);
  const serviceMap = renderService(apiDoc, { operationIdSplitter: cfg.operationIdSplitter });

  // render template
  let content = ejs.render(fs.readFileSync(template, { encoding: 'utf-8' }), { serviceMap });
  if (banner) content = banner + '\n' + content;

  // pretty
  const prettierOpt = await prettier.resolveConfig(process.cwd());
  content = prettier.format(content, {
    parser: 'typescript',
    trailingComma: 'es5',
    tabWidth: 2,
    semi: true,
    singleQuote: true,
    jsxSingleQuote: true,
    printWidth: 100,
    bracketSpacing: true,
    arrowParens: 'avoid',
    ...prettierOpt,
  });

  // dump
  if (dump) fs.writeFileSync(dump, content, { encoding: 'utf-8' });

  return content;
};
