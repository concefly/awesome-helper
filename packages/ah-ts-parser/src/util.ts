import { inspect } from 'util';

export const log = (...args: any[]) => {
  console.log(...args.map(arg => inspect(arg, { showHidden: false, depth: null })));
};
