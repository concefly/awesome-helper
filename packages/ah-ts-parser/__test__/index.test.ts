import { Ts2Schema } from '../src';
import { inspect } from 'util';

export const log = (...args: any[]) => {
  console.log(...args.map(arg => inspect(arg, { showHidden: false, depth: null })));
};

describe('index', () => {
  it('normal', () => {
    const ts2Schema = new Ts2Schema('./tsconfig.json');
    const s = ts2Schema.getSchema('**/fixture.ts');
    expect(s).toMatchSnapshot();
  });
});
