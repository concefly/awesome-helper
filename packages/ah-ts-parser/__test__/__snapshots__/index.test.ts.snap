// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index normal 1`] = `
Object {
  "properties": Object {
    "files": Object {
      "items": Array [
        Object {
          "properties": Object {
            "classes": Object {
              "items": Array [
                Object {
                  "properties": Object {
                    "flag": Object {
                      "title": undefined,
                      "type": "string",
                    },
                    "foo": Object {
                      "format": "function",
                      "properties": Object {
                        "signatures": Object {
                          "items": Array [
                            Object {
                              "properties": Object {
                                "params": Object {
                                  "items": Array [
                                    Object {
                                      "title": undefined,
                                      "type": "string",
                                    },
                                    Object {
                                      "title": undefined,
                                      "type": "integer",
                                    },
                                    Object {
                                      "properties": Object {
                                        "city": Object {
                                          "title": undefined,
                                          "type": "string",
                                        },
                                      },
                                      "type": "object",
                                    },
                                  ],
                                  "type": "array",
                                },
                                "return": Object {
                                  "title": undefined,
                                  "type": "string",
                                },
                              },
                              "type": "object",
                            },
                          ],
                          "type": "array",
                        },
                      },
                      "title": "foo",
                      "type": "object",
                    },
                    "foo2": Object {
                      "format": "function",
                      "properties": Object {
                        "signatures": Object {
                          "items": Array [
                            Object {
                              "properties": Object {
                                "params": Object {
                                  "items": Array [],
                                  "type": "array",
                                },
                                "return": Object {
                                  "title": undefined,
                                  "type": "integer",
                                },
                              },
                              "type": "object",
                            },
                          ],
                          "type": "array",
                        },
                      },
                      "title": "foo2",
                      "type": "object",
                    },
                    "foo3": Object {
                      "format": "function",
                      "properties": Object {
                        "signatures": Object {
                          "items": Array [
                            Object {
                              "properties": Object {
                                "params": Object {
                                  "items": Array [],
                                  "type": "array",
                                },
                                "return": Object {
                                  "properties": Object {
                                    "data": Object {
                                      "title": undefined,
                                      "type": "string",
                                    },
                                  },
                                  "type": "object",
                                },
                              },
                              "type": "object",
                            },
                          ],
                          "type": "array",
                        },
                      },
                      "title": "foo3",
                      "type": "object",
                    },
                    "foo4": Object {
                      "format": "function",
                      "properties": Object {
                        "signatures": Object {
                          "items": Array [
                            Object {
                              "properties": Object {
                                "params": Object {
                                  "items": Array [],
                                  "type": "array",
                                },
                                "return": Object {
                                  "properties": Object {
                                    "foo3Result": Object {
                                      "properties": Object {
                                        "data": Object {
                                          "title": undefined,
                                          "type": "string",
                                        },
                                      },
                                      "type": "object",
                                    },
                                  },
                                  "type": "object",
                                },
                              },
                              "type": "object",
                            },
                          ],
                          "type": "array",
                        },
                      },
                      "title": "foo4",
                      "type": "object",
                    },
                  },
                  "title": "Animal",
                  "type": "object",
                },
                Object {
                  "properties": Object {
                    "foo5": Object {
                      "format": "function",
                      "properties": Object {
                        "signatures": Object {
                          "items": Array [
                            Object {
                              "properties": Object {
                                "params": Object {
                                  "items": Array [],
                                  "type": "array",
                                },
                                "return": Object {
                                  "properties": Object {
                                    "data": Object {
                                      "properties": Object {
                                        "foo3Result": Object {
                                          "properties": Object {
                                            "data": Object {
                                              "title": undefined,
                                              "type": "string",
                                            },
                                          },
                                          "type": "object",
                                        },
                                      },
                                      "type": "object",
                                    },
                                  },
                                  "type": "object",
                                },
                              },
                              "type": "object",
                            },
                          ],
                          "type": "array",
                        },
                      },
                      "title": "foo5",
                      "type": "object",
                    },
                    "walk": Object {
                      "format": "function",
                      "properties": Object {
                        "decorators": Object {
                          "items": Array [
                            Object {
                              "properties": Object {
                                "arguments": Object {
                                  "items": Array [
                                    Object {
                                      "const": "'meta'",
                                      "type": "string",
                                    },
                                  ],
                                  "type": "array",
                                },
                              },
                              "title": "addMeta",
                              "type": "object",
                            },
                          ],
                          "type": "array",
                        },
                        "signatures": Object {
                          "items": Array [
                            Object {
                              "properties": Object {
                                "params": Object {
                                  "items": Array [],
                                  "type": "array",
                                },
                                "return": Object {
                                  "format": "promise",
                                  "properties": Object {
                                    "typeArguments": Object {
                                      "items": Array [
                                        Object {
                                          "title": undefined,
                                          "type": "integer",
                                        },
                                      ],
                                      "type": "array",
                                    },
                                  },
                                  "type": "object",
                                },
                              },
                              "type": "object",
                            },
                          ],
                          "type": "array",
                        },
                      },
                      "title": "walk",
                      "type": "object",
                    },
                  },
                  "title": "Cat",
                  "type": "object",
                },
              ],
              "type": "array",
            },
            "functions": Object {
              "items": Array [
                Object {
                  "format": "function",
                  "properties": Object {
                    "signatures": Object {
                      "items": Array [
                        Object {
                          "properties": Object {
                            "params": Object {
                              "items": Array [
                                Object {
                                  "title": undefined,
                                  "type": "string",
                                },
                              ],
                              "type": "array",
                            },
                            "return": Object {
                              "format": "function",
                              "properties": Object {
                                "signatures": Object {
                                  "items": Array [
                                    Object {
                                      "properties": Object {
                                        "params": Object {
                                          "items": Array [
                                            Object {},
                                            Object {
                                              "title": undefined,
                                              "type": "string",
                                            },
                                            Object {
                                              "properties": Object {
                                                "configurable": Object {},
                                                "enumerable": Object {},
                                                "get": Object {},
                                                "set": Object {},
                                                "value": Object {},
                                                "writable": Object {},
                                              },
                                              "type": "object",
                                            },
                                          ],
                                          "type": "array",
                                        },
                                        "return": Object {},
                                      },
                                      "type": "object",
                                    },
                                  ],
                                  "type": "array",
                                },
                              },
                              "title": "__function",
                              "type": "object",
                            },
                          },
                          "type": "object",
                        },
                      ],
                      "type": "array",
                    },
                  },
                  "title": "addMeta",
                  "type": "object",
                },
                Object {
                  "format": "function",
                  "properties": Object {
                    "signatures": Object {
                      "items": Array [
                        Object {
                          "properties": Object {
                            "params": Object {
                              "items": Array [
                                Object {
                                  "title": undefined,
                                  "type": "string",
                                },
                                Object {
                                  "title": undefined,
                                  "type": "integer",
                                },
                                Object {
                                  "properties": Object {
                                    "city": Object {
                                      "title": undefined,
                                      "type": "string",
                                    },
                                  },
                                  "type": "object",
                                },
                              ],
                              "type": "array",
                            },
                            "return": Object {
                              "properties": Object {
                                "age": Object {
                                  "title": undefined,
                                  "type": "string",
                                },
                                "list": Object {},
                                "name": Object {
                                  "title": undefined,
                                  "type": "string",
                                },
                              },
                              "type": "object",
                            },
                          },
                          "type": "object",
                        },
                      ],
                      "type": "array",
                    },
                  },
                  "title": "foo",
                  "type": "object",
                },
              ],
              "type": "array",
            },
          },
          "title": "__test__/fixture.ts",
          "type": "object",
        },
      ],
      "type": "array",
    },
  },
  "type": "object",
}
`;
